Content.Shared/Weapons/Reflect/SharedReflectSystem.cs

```diff
@@ -0,0 +1,87 @@
+using Content.Shared.Audio;
+using Content.Shared.Popups;
+using Content.Shared.Projectiles;
+using Robust.Shared.Random;
+using Robust.Shared.Physics.Systems;
+using Content.Shared.Weapons.Ranged;
+using Content.Shared.Hands.Components;
+using Robust.Shared.GameStates;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Weapons.Reflect;
+
+/// <summary>
+/// This handles reflecting projectiles and hitscan shots.
+/// </summary>
+public abstract class SharedReflectSystem : EntitySystem
+{
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+    [Dependency] private readonly SharedProjectileSystem _projectile = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<SharedHandsComponent, PreventProjectileCollideEvent>(TryReflectProjectile);
+        SubscribeLocalEvent<SharedHandsComponent, HitScanReflectAttempt>(TryReflectHitScan);
+        SubscribeLocalEvent<ReflectComponent, ComponentHandleState>(OnHandleState);
+        SubscribeLocalEvent<ReflectComponent, ComponentGetState>(OnGetState);
+    }
+
+    private static void OnHandleState(EntityUid uid, ReflectComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not ReflectComponentState state) return;
+        component.Enabled = state.Enabled;
+        component.Chance = state.Chance;
+        component.Spread = state.Spread;
+    }
+
+    private static void OnGetState(EntityUid uid, ReflectComponent component, ref ComponentGetState args)
+    {
+        args.State = new ReflectComponentState(component.Enabled, component.Chance, component.Spread);
+    }
+
+    private void TryReflectProjectile(EntityUid uid, SharedHandsComponent hands, ref PreventProjectileCollideEvent args)
+    {
+        foreach (var (_, hand) in hands.Hands)
+        {
+            if (TryComp<ReflectComponent>(hand.HeldEntity, out var reflect) &&
+                reflect.Enabled && 
+                _random.Prob(reflect.Chance))
```

Comment:
This should be checking for active hand as otherwise they can freely shoot you with no recourse.